src/plots/cartesian/layout_defaults.js

```diff
@@ -93,6 +99,80 @@ module.exports = function supplyLayoutDefaults(layoutIn, layoutOut, fullData) {
         }
     }
 
+    function getXname() {
+        return (trace.xaxis) ? id2name(trace.xaxis) : undefined; // TODO: why these should be left undefined?
+    }
+
+    function getYname() {
+        return (trace.yaxis) ? id2name(trace.yaxis) : undefined; // TODO: why these should be left undefined?
+    }
+
+    function includesOnly(desiredType, xaNameIn, yaNameIn) {
+        var result = false;
+        for(var k = 0; k < fullData.length; k++) {
```

Comment:
Hmm. Am I seeing this right, we could `O(fullData.length^2)` loops? Is that really necessary to detect if all traces on a given axis are `funnel` traces. Am I missing something?

Reply:
Good call. Reworked in https://github.com/plotly/plotly.js/pull/3817/commits/a650170bfcee35743e0b3eb5aa42e2e12b8518d6.

Reply:
Wow. So clean. Thanks!